rust/main/agents/validator/src/submit.rs

```diff
@@ -313,3 +332,321 @@ impl ValidatorSubmitterMetrics {
+        );
+        assert_eq!(
+            reorg_event.checkpoint_index,
+            expected_local_merkle_tree.index()
+        );
+        // timestamp diff should be less than 1 second
+        let timestamp_diff = reorg_event.unix_timestamp as i64 - unix_timestamp as i64;
+        assert!(timestamp_diff.abs() < 1);
+
+        assert_eq!(reorg_event.reorg_period, expected_reorg_period);
+    }
+
+    #[tokio::test]
+    #[should_panic(expected = "Incorrect tree root, something went wrong.")]
+    async fn reorg_is_detected_and_persisted_to_checkpoint_storage() {
+        let unix_timestamp = chrono::Utc::now().timestamp() as u64;
+        let expected_reorg_period = 12;
+
+        let pre_reorg_merke_insertions = vec![
+            MerkleTreeInsertion::new(0, H256::random()),
+            MerkleTreeInsertion::new(1, H256::random()),
+            MerkleTreeInsertion::new(2, H256::random()),
+        ];
+        let mut expected_local_merkle_tree = IncrementalMerkle::default();
+        for insertion in pre_reorg_merke_insertions.iter() {
+            expected_local_merkle_tree.ingest(insertion.message_id());
+        }
+
+        // the last leaf is different post-reorg
+        let post_reorg_merkle_insertions = vec![
+            pre_reorg_merke_insertions[0].clone(),
+            pre_reorg_merke_insertions[1].clone(),
+            MerkleTreeInsertion::new(2, H256::random()),
+        ];
+        let mut mock_onchain_merkle_tree = IncrementalMerkle::default();
+        for insertion in post_reorg_merkle_insertions.iter() {
+            mock_onchain_merkle_tree.ingest(insertion.message_id());
+        }
+
+        // assert the reorg resulted in different merkle tree roots
+        assert_ne!(
+            mock_onchain_merkle_tree.root(),
+            expected_local_merkle_tree.root()
+        );
+
+        // the db returns the pre-reorg merkle tree insertions
+        let mut db = MockDb::new();
+        db.expect_retrieve_merkle_tree_insertion_by_leaf_index()
+            .returning(move |sequence| {
+                Ok(Some(pre_reorg_merke_insertions[*sequence as usize].clone()))
+            });
+
+        // boilerplate mocks
+        let mut mock_merkle_tree_hook = MockMerkleTreeHook::new();
+        mock_merkle_tree_hook
+            .expect_address()
+            .returning(|| H256::from_low_u64_be(0));
+        let dummy_domain = dummy_domain(0, "dummy_domain");
+        mock_merkle_tree_hook
+            .expect_domain()
+            .return_const(dummy_domain.clone());
+
+        // expect the checkpoint syncer to post the reorg event to the checkpoint storage
+        // and not submit any checkpoints (this is checked implicitly, by not setting any `expect`s)
+        let mut mock_checkpoint_syncer = MockCheckpointSyncer::new();
+        let mock_onchain_merkle_tree_clone = mock_onchain_merkle_tree.clone();
+        mock_checkpoint_syncer
+            .expect_write_reorg_status()
+            .once()
```

Comment:
does a panic occur if this function is never called?

Reply:
it does, this requires the function to be called exactly once: https://docs.rs/mockall/latest/mockall/examples/__mock_MockBoo_Bah/__bah/struct.Expectation.html#method.once